File: Python/JDI/web/selenium/elements/base/base_element.py
from JDI.web.selenium.elements.api_interact.get_element_module import GetElementModule


class BaseElement(object):
    name = None
    parent = None
    avatar = None

    def __init__(self, by_locator=None):
        self.avatar = GetElementModule(by_locator, self)

    def get_driver(self):
        return self.avatar.get_driver()

    def __str__(self):
        return self.__class__.__name__

    def _get_type_name(self):
        return self.__class__.__name__

    def get_name(self):
        return self.name if self.name is not None else self._get_type_name()

    def get_parent(self):
        return self.parent

    def init(self, parent, avatar):
        from JDI.web.selenium.elements.cascade_init import WebCascadeInit
        WebCascadeInit.init_elements(self)
        self.set_avatar(avatar)
        self.set_parent(parent)
        return self

    def set_avatar(self, avatar):
        self.avatar = avatar
        return self

    def set_parent(self, parent):
        self.parent = parent

    def get_locator(self):
        return self.avatar.by_locator

    def __str__(self):
        str = "Name " + self.__class__.__name__
        if "by_locator" in dir(self.avatar):
            if self.avatar.by_locator is not None:
                str += "; Locator: %s:'%s'" % (self.avatar.by_locator[0], self.avatar.by_locator[1])
        if self.parent is not None:
            if "avatar" in dir(self.parent):
                if self.parent.avatar.by_locator is not None:
                    str += "; Parent: %s:'%s'" % (self.parent.avatar.by_locator[0], self.parent.avatar.by_locator[1])

        return str

    def has_locator(self):
        return self.avatar.has_locator()

File: Java/JDI/jdi-uitest-win/src/main/java/com/epam/jdi/uitests/win/winnium/elements/BaseElement.java
package com.epam.jdi.uitests.win.winnium.elements;

import com.epam.jdi.uitests.core.annotations.AnnotationsUtil;
import com.epam.jdi.uitests.core.annotations.functions.Functions;
import com.epam.jdi.uitests.core.interfaces.base.IAvatar;
import com.epam.jdi.uitests.core.interfaces.base.IBaseElement;
import com.epam.jdi.uitests.core.settings.JDISettings;
import com.epam.jdi.uitests.win.winnium.actions.ActionInvoker;
import com.epam.jdi.uitests.win.winnium.elements.apiInteract.GetElementModule;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

import java.lang.reflect.Field;
import java.text.MessageFormat;

import static com.epam.jdi.uitests.core.settings.JDISettings.*;
import static java.util.concurrent.TimeUnit.SECONDS;

public abstract class BaseElement implements IBaseElement {
    private Object parent;
    private String name;
    private String varName;
    private String typeName;

    protected GetElementModule avatar;
    protected ActionInvoker invoker = new ActionInvoker();

    public BaseElement() {
        avatar = new GetElementModule(this);
    }

    public By getLocator() {
        return avatar.getByLocator();
    }
    public void setLocator(By locator) {
        avatar.setByLocator(locator);
    }

    @Override
    public GetElementModule getAvatar() {
        return avatar;
    }

    public WebDriver getDriver() {
        return getAvatar().getDriver();
    }

    public void setWaitTimeout(int seconds) {
        logger.debug("Set wait timeout to " + seconds);
        getDriver().manage().timeouts().implicitlyWait(seconds, SECONDS);
        timeouts.setCurrentTimeoutSec(seconds);
    }

    public void restoreWaitTimeout() {
        setWaitTimeout(timeouts.getDefaultTimeoutSec());
    }

    @Override
    public void setParent(Object parent) {
        this.parent = parent;
    }

    @Override
    public void setName(Field field) {
        this.name = AnnotationsUtil.getElementName(field);
        this.varName = field.getName();
    }

    @Override
    public Object getParent() {
        return parent;
    }

    @Override
    public void setTypeName(String typeName) {
        this.typeName = typeName;
    }

    @Override
    public String getName() {
        return name != null ? name : getTypeName();
    }

    @Override
    public String printContext() {
        if (!(parent instanceof BaseElement))
            return "";

        BaseElement parentBaseElement = (BaseElement)parent;
        By locator = parentBaseElement.avatar.getByLocator();
        if (locator == null)
            return "";

        String parentContext = parentBaseElement.printContext();
        if (parentContext.equals(""))
            return locator.toString();

        return locator.toString() + "; " + parentContext;
    }

    @Override
    public String toString() {
        return MessageFormat.format(shortLogMessagesFormat
                        ? "{1} ''{0}'' ({2}.{3}; {4})"
                        : "Name: ''{0}'', Type: ''{1}'' In: ''{2}'', {4}",
                getName(), getTypeName(), getParentName(), getVarName(), avatar);
    }

    private String getParentName() {
        return parent == null ? "" : parent.getClass().getSimpleName();
    }

    private String getTypeName() {
        return (typeName != null) ? typeName : getClass().getSimpleName();
    }

    private String getVarName() {
        return varName != null ? varName : getName();
    }

    public void setAvatar(GetElementModule avatar) {
        this.avatar = avatar;
    }

    public BaseElement setAvatar(IAvatar avatar) {
        this.avatar = (GetElementModule) avatar.copy();
        return this;
    }

    public boolean hasLocator() {
        return avatar.hasLocator();
    }

    public ActionInvoker getInvoker() {
        return invoker;
    }

    @Override
    public void setFunction(Functions function) {
        if (function != Functions.NONE)
            throw JDISettings.exception("Not supported");
    }
}
